feat: usePopper Hook

FezVrasta opened the pull request:
```js
// @flow
import * as React from 'react';
import { usePopper } from 'react-popper';

const App = () => {
  const [referenceElement, setReferenceElement] = React.useState<?Element>(null);
  const [popperElement, setPopperElement] = React.useState<?HTMLElement>(null);
  const [arrowElement, setArrowElement] = React.useState<?HTMLElement>(null);

  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    modifiers: [
      { name: 'arrow', options: { element: arrowElement } },
    ],
  });

  return (
    <>
      <button ref={setReferenceElement}>Reference Element</button>
      <div ref={setPopperElement} style={styles.popper} {...attributes.popper}>
        Popper Element
        <div ref={setArrowElement} style={styles.arrow} />
      </div>
    </>
  );
}
```

CodeSandbox demo:
https://codesandbox.io/s/cold-night-u1oji

atomiks commented:
Can you show an example/demo of this hook in practice?

FezVrasta commented:
@atomiks I added an example on the first post

bengry commented:
1. Why is this example not using `React.useRef` and instead uses `React.useState`?
2. Maybe do something similar to downshift in regards to the props-to-be-spread API, something like this:

```js
// @flow
import * as React from "react";
import { usePopper } from "react-popper";

const App = () => {
  const [referenceElement, setReferenceElement] = React.useState<?Element>(
    null
  );
  const [popperElement, setPopperElement] = React.useState<?HTMLElement>(null);
  const [arrowElement, setArrowElement] = React.useState<?HTMLElement>(null);

  const { styles, attributes } = usePopper(referenceElement, popperElement, {
    modifiers: [{ name: "arrow", options: { element: arrowElement } }],
  });

  return (
    <>
      <button ref={setReferenceElement}>Reference Element</button>
      <div
        ref={setPopperElement}
        // If you want to specify additional props (e.g. style) - pass them in when calling `getPopperProps()`. Otherwise this is optional.
        // Also, no need to for the `styles`
        {...attributes.getPopperProps({ style: { background: "red" } })}
      >
        Popper Element
        <div ref={setArrowElement} style={styles.arrow} />
      </div>
    </>
  );
};

```

FezVrasta commented:
> 1. Why is this example not using `React.useRef` and instead uses `React.useState`?

https://github.com/popperjs/react-popper/issues/241#issuecomment-591411605

> 2. Maybe do something similar to downshift in regards to the props-to-be-spread API, something like this [...]

Merging props wouldn't add a lot in our case, the values are all strings, so doing `{ ...yourStyles, ...styles.popper }` would do exactly the same.

bengry commented:
> > 1. Maybe do something similar to downshift in regards to the props-to-be-spread API, something like this [...]
> 
> Merging props wouldn't add a lot in our case, the values are all strings, so doing `{ ...yourStyles, ...styles.popper }` would do exactly the same.

Right, but it allows adding additional inputs in the future, should any be added [without breaking the API], and also - it means you only need to hook one thing up to use popper (call `getPopperProps()`), not two - (spread `attributes.popper` + passing in `styles.popper`).

FezVrasta commented:
attributes and styles are separated because I want to let the user decide what to apply

bengry commented:
@FezVrasta I've updated my ealier comment with how to solve this, while keeping the [minimal usable] API lean (since passing it the popper styles is required for the `div` to be positioned correctly).

> ```js
> <div {...attributes.popper} />
> 
> // vs.
> 
> <div {...attributes.popper} styles={{...attributes.popper.styles, { background: 'red' }} />
> ```

FezVrasta commented:
> ```
> <div {...attributes.popper} styles={{...attributes.popper.styles, { background: 'red' }} />
> ```

Isn't the above example basically the same number of characters as the following?

```js
<div {...attributes} styles={{ ...styles.popper, background: 'red' }} />
```

bengry commented:
Is it, but that's the uncommon case I think. The more common use would be: `<div {...attributes.popper} />`. I don't find myself passing an inline `style` to the popper node very often (ever?). Even if I need to style it, I'd usually use `className`.

FezVrasta commented:
The reasoning behind this data structure is that it mimics the one exposed by Popper, so it doesn't add an additional abstraction layer on top of it. As I explained in different occasions, react-popper is meant to be an as thin as possible layer on top of Popper, for more complex use-cases I'd leverage on wrappers on top of it (like react-tippy)

atomiks commented on src/usePopper.js:
```diff
@@ -0,0 +1,100 @@
+        });
+      },
+      requires: ['computeStyles'],
+      options: { setState },
```
What's the purpose of this?

FezVrasta replied:
Nothing, I have to remove it, lol

atomiks commented:
Looks great imo

atomiks commented on src/usePopper.js:
```diff
@@ -0,0 +1,99 @@
+    [setState]
+  );
+
+  useLayoutEffect(() => {
```
`useLayoutEffect` needs to check browser env check to avoid the SSR warning

FezVrasta replied:
🤦‍♂️

atomiks replied:
pinging this again

FezVrasta replied:
I'm wondering if we could follow the React team advice where they suggest not to run the code at all on server-side?

atomiks commented on src/usePopper.js:
```diff
@@ -0,0 +1,99 @@
+        const elements = Object.keys(state.elements);
+
+        setState({
+          styles: Object.fromEntries(
```
I don't think we should use `fromEntries` due to old browsers, keep polyfills as minimal as `@popperjs/core`

FezVrasta replied:
I doubt it's going to be an issue, those React libraries are usually consumed through webpack+babel

atomiks commented on src/usePopper.js:
```diff
@@ -0,0 +1,99 @@
+    attributes: {},
+  });
+  const createPopper = useMemo(
+    () => options.createPopper || defaultCreatePopper,
```
Does `defaultCreatePopper` get tree-shaken?

FezVrasta replied:
No, but at least users can pass their own `createPopper` function if needed

atomiks commented on src/usePopper.js:
```diff
@@ -0,0 +1,99 @@
+  const popperInstanceRef = useRef();
+  const [state, setState] = useState<State>({
+    styles: {
+      popper: { position: 'absolute', left: '0', top: '0' },
```
`position: options.strategy` (https://github.com/popperjs/popper-core/pull/1055)

FezVrasta commented:
I pushed some work to convert the render-props implementation to use hooks, it's half broken, I'd appreciate some help to get it working.

You can see the issues by running `yarn demo:dev`

FezVrasta commented:
Ok so we are almost there, there's just one problem, the poppers are not positioned (they default to `top: 0, left: 0`).

@atomiks do you have time to take a look? I'm not quite sure what's going on 😓

atomiks commented:
It's a stale dependency problem w/ `useMemo`

This fixes it:

```js
  const childrenProps = React.useMemo(
    () => ({
      ref: setPopperElement,
      style: styles.popper,
      placement: state ? state.placement : placement,
      hasPopperEscaped: state
        // need optional ?. here
        ? state.modifiersData.hide?.hasPopperEscaped
        : null,
      isReferenceHidden: state
        // and here
        ? state.modifiersData.hide?.isReferenceHidden
        : null,
      arrowProps: {
        style: styles.arrow,
        ref: setArrowElement,
      },
      forceUpdate: forceUpdate || NOOP,
      update: update || NOOP_PROMISE,
    }),
    [setPopperElement, state]
  );
```

FezVrasta commented:
Thanks! I added the eslint rule to make sure it doesn't happen again.

FezVrasta commented:
So, I suppose the PR is ready for a final review.

FezVrasta commented:
Tests fixed, I replaced Enzyme with react-testing-library since it seems to work better with modern React versions. I also added tests for the new hook.

Imballinst commented:
hi @FezVrasta, thank you for this amazing work! Question: why does in the CodeSandbox demo, the `div` for arrow is located below the popper instead of above the popper? Is this a bug from https://github.com/popperjs/popper-core?

I also face the same thing in my local, but I don't know whether it's because of this lib or because of upstream repository. I also don't quite know whether I'm configuring something wrong, though I copy-pasted exactly the same code like in the documentation for `usePopper`. Thanks again!

atomiks commented:
@Imballinst Popper only positions the arrow along a single axis in order to center it to the reference. For vertical placements it's the `x` axis. It's your responsibility to position it on the other axis in CSS.

FezVrasta commented:
I’m sorry I’m from mobile so I can’t look into it, but  make sure you are applying the correct CSS to position the arrow. Popper provides only the offset to center it. You can look at the popper.js.org tutorial

Imballinst commented:
@atomiks @FezVrasta Ah, okay! Got it. That should be enough as a clue for me. Thanks and sorry for the noise. Have a nice weekend!

pleerock commented:
shouldn't `useRef` be used instead of `useState` for element references?

FezVrasta commented:
`useRef` allows to react when the reference element changes identity.

yuriybelike commented:
Is the demo's 

> // FIXME: we probably need some logic in the hook to detect when the provided elements are `null`

still applicable? I'm having some issues setting `referenceElement` to null in my code

FezVrasta commented:
> Is the demo's 
> 
> 
> 
> > // FIXME: we probably need some logic in the hook to detect when the provided elements are `null`
> 
> 
> 
> still applicable? I'm having some issues setting `referenceElement` to null in my code.

It should work with null elements

sarathms commented:
The codesanbox example linked above doesn't seem to work anymore. This is what I see:
![ - 43](https://user-images.githubusercontent.com/700829/87565691-fb7b3480-c68f-11ea-828b-1abc49bda312.png)

How to get it to work?

FezVrasta commented:
It's working, it's just a silly example where the popper sits below the button 😅
